### ISSUE TITLE
How to use the RAII concept to sync data between device and host?

### ISSUE THREAD

**Issue by varunhiremath** [kokkos/kokkos#2546]
Hello,

I'm experimenting with using Kokkos to run some simple loops on GPU involving std::vector data. Please see the below snippet of code which contains three functions:
1) testSerial: a simple serial loop test over vector data
2) testDirectKokkos: same test implemented using Kokkos (standard way using Kokkos::View)
3) testKokkosField: same test implemented using a wrapper KokkosField class to hide most of the Kokkos specific details

```
#include <vector>
#include <iostream>
#include <numeric>

#include "Kokkos_Core.hpp"

int const size = 1e7;

void testSerial()
{
  std::vector<float> A(size,0.0);
  std::vector<float> B(size,0.0);
  std::vector<float> C(size,0.0);

  for (int i = 0; i < size; ++i)
  {
    A[i] = i;
    B[i] = i;
  }

  Kokkos::Timer kTimer;
  for (int i = 0; i < size; ++i)
  {
    C[i] = A[i]*B[i];
  }
  float const time = kTimer.seconds();

  float const result = std::accumulate(C.begin(), C.end(), 0.0);

  std::cout<<"testSerial: result = " << result << " in time = " << time << "(s)" << std::endl;
}

void testDirectKokkos()
{
  std::vector<float> A(size,0.0);
  std::vector<float> B(size,0.0);
  std::vector<float> C(size,0.0);

  for (int i = 0; i < size; ++i)
  {
    A[i] = i;
    B[i] = i;
  }

  typedef Kokkos::View<float*,Kokkos::Cuda> GpuView;
  GpuView gpuA("gpuA", size);
  GpuView gpuB("gpuB", size);
  GpuView gpuC("gpuC", size);

  typename GpuView::HostMirror hostX = Kokkos::create_mirror_view(gpuA);

  for (int i = 0; i < size; ++i)
    hostX(i) = A[i];
  Kokkos::deep_copy(gpuA, hostX);

  for (int i = 0; i < size; ++i)
    hostX(i) = B[i];
  Kokkos::deep_copy(gpuB, hostX);

  Kokkos::Timer kTimer;
  Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::Cuda>(0,size), KOKKOS_LAMBDA(int i)
  {
    gpuC(i) = gpuA(i)*gpuB(i);
  });
  float const time = kTimer.seconds();

  Kokkos::deep_copy(hostX, gpuC);
  for (int i = 0; i < size; ++i)
    C[i] = hostX(i);

  float const result = std::accumulate(C.begin(), C.end(), 0.0);

  std::cout<<"testDirectKokkos: result = " << result << " in time = " << time << "(s)" << std::endl;
}

// A wrapper class for the Kokkos data
class KokkosField
{
public:
  KokkosField(std::vector<float> &vec)
    : _data(vec)
    , _size(vec.size())
    , _gpuField("_gpuField", _size)
    , _hostX(Kokkos::create_mirror_view(_gpuField))
  {
    for (int i = 0; i < _size; ++i)
      _hostX(i) = _data[i];
    Kokkos::deep_copy(_gpuField, _hostX);
  }

  ~KokkosField()
  {
    Kokkos::deep_copy(_hostX, _gpuField);
    for (int i = 0; i < _size; ++i)
      _data[i] = _hostX(i);
  }

  KOKKOS_FORCEINLINE_FUNCTION float& operator[](int const i) const
  {
    return _gpuField(i);
  }

private:
  std::vector<float> &_data;
  int const _size;
  typedef Kokkos::View<float*,Kokkos::Cuda> GpuView;
  GpuView _gpuField;
  typename GpuView::HostMirror _hostX;
};

void testKokkosField()
{
  std::vector<float> A(size,0.0);
  std::vector<float> B(size,0.0);
  std::vector<float> C(size,0.0);

  for (int i = 0; i < size; ++i)
  {
    A[i] = i;
    B[i] = i;
  }

  float time;
  {
    KokkosField kA(A);
    KokkosField kB(B);
    KokkosField kC(C);

    Kokkos::Timer kTimer;
    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::Cuda>(0,size), KOKKOS_LAMBDA(int i)
    {
      kC[i] = kA[i]*kB[i];
    });
    time = kTimer.seconds();
  }

  float const result = std::accumulate(C.begin(), C.end(), 0.0);
  std::cout<<"testKokkosField: result = " << result << " in time = " << time << "(s)" << std::endl;
}

int main()
{
  Kokkos::InitArguments args;
  args.num_threads = 8;
  args.num_numa    = 2;
  args.device_id   = 0;
  Kokkos::initialize(args);

  testSerial();
  testDirectKokkos();
  testKokkosField();

  Kokkos::finalize();
  return 1;
}
```

If you run the above code you will find that testDirectKokkos runs much faster than testSerial, but testKokkosField runs much slower. I've figured this is due to the private data member `std::vector<float> &_data;` in the KokkosField wrapper class which gets captured by value by the lambda function and slows everything down (as explained [here](https://github.com/kokkos/kokkos/wiki/Lambda-Dispatch#lambdas-inside-classes-with-cuda)). The below patch fixes the issue but makes the code a bit ugly as I now need to explicitly copy the data back. I was just wondering if there is any workaround for this? Is there a way to mark some class member as host only so it doesn't get copied to GPU?

```
// patch to fix the above issue
@@ -78,21 +78,22 @@
 {
 public:
   KokkosField(std::vector<float> &vec)
-    : _data(vec)
-    , _size(vec.size())
+  //    : _data(vec)
+    : _size(vec.size())
     , _gpuField("_gpuField", _size)
     , _hostX(Kokkos::create_mirror_view(_gpuField))
   {
     for (int i = 0; i < _size; ++i)
-      _hostX(i) = _data[i];
+      _hostX(i) = vec[i];
     Kokkos::deep_copy(_gpuField, _hostX);
   }
 
-  ~KokkosField()
+  //~KokkosField()
+  void syncData(std::vector<float> &data)
   {
     Kokkos::deep_copy(_hostX, _gpuField);
     for (int i = 0; i < _size; ++i)
-      _data[i] = _hostX(i);
+      data[i] = _hostX(i);
   }
 
   KOKKOS_FORCEINLINE_FUNCTION float& operator[](int const i) const
@@ -101,7 +102,7 @@
   }
 
 private:
-  std::vector<float> &_data;
+  //std::vector<float> &_data;
   int const _size;
   typedef Kokkos::View<float*,Kokkos::Cuda> GpuView;
   GpuView _gpuField;
@@ -132,6 +133,7 @@
       kC[i] = kA[i]*kB[i];
     });
     time = kTimer.seconds();
+    kC.syncData(C);
   }
 
   float const result = std::accumulate(C.begin(), C.end(), 0.0);
```

Thanks,
Varun

**Comment by dalg24** [comment]
Try this
```diff
@@ -49,13 +49,9 @@

   typename GpuView::HostMirror hostX = Kokkos::create_mirror_view(gpuA);

-  for (int i = 0; i < size; ++i)
-    hostX(i) = A[i];
-  Kokkos::deep_copy(gpuA, hostX);
-
-  for (int i = 0; i < size; ++i)
-    hostX(i) = B[i];
-  Kokkos::deep_copy(gpuB, hostX);
+  typedef Kokkos::View<float*,Kokkos::HostSpace,Kokkos::MemoryTraits<Kokkos::Unmanaged>> UnmanagedHostView;
+  Kokkos::deep_copy(gpuA, UnmanagedHostView(A.data(), A.size()));
+  Kokkos::deep_copy(gpuB, UnmanagedHostView(B.data(), B.size()));

   Kokkos::Timer kTimer;
   Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::Cuda>(0,size), KOKKOS_LAMBDA(int i)
@@ -64,9 +60,7 @@
   });
   float const time = kTimer.seconds();

-  Kokkos::deep_copy(hostX, gpuC);
-  for (int i = 0; i < size; ++i)
-    C[i] = hostX(i);
+  Kokkos::deep_copy(UnmanagedHostView(C.data(), C.size()), gpuC);

   float const result = std::accumulate(C.begin(), C.end(), 0.0);
```

**Comment by dalg24** [comment]
Sure you could change the definition of `KokkosField` to
```diff
@@ -78,21 +72,16 @@
 {
 public:
   KokkosField(std::vector<float> &vec)
-    : _data(vec)
-    , _size(vec.size())
+    : _size(vec.size())
     , _gpuField("_gpuField", _size)
-    , _hostX(Kokkos::create_mirror_view(_gpuField))
+    , _hostX(UnmanagedHostView(vec.data(), vec.size()))
   {
-    for (int i = 0; i < _size; ++i)
-      _hostX(i) = _data[i];
     Kokkos::deep_copy(_gpuField, _hostX);
   }

   ~KokkosField()
   {
     Kokkos::deep_copy(_hostX, _gpuField);
-    for (int i = 0; i < _size; ++i)
-      _data[i] = _hostX(i);
   }

   KOKKOS_FORCEINLINE_FUNCTION float& operator[](int const i) const
@@ -101,8 +90,8 @@
   }

 private:
-  std::vector<float> &_data;
   int const _size;
+  typedef Kokkos::View<float*,Kokkos::HostSpace,Kokkos::MemoryTraits<Kokkos::Unmanaged>> UnmanagedHostView;
   typedef Kokkos::View<float*,Kokkos::Cuda> GpuView;
   GpuView _gpuField;
   typename GpuView::HostMirror _hostX;
```
but your "fields" are captured by value and the destructor will be called a bunch of times (i.e. you copy the same data back from the device to the host multiple times).  You can add print statements in  the constructor and the destructor if you want to see.

Also another comment:  you omitted the data transfer in `testDirectKokkos()` so you are comparing apples and oranges ;)

**Comment by varunhiremath** [comment]
Hi Damien,

Thank you very much for pointing me in the right direction! Yes I realize that I should only copy data for non-const fields that are changing. I think the below code does the trick for me. I override the copy constructor to disable copying of data in the destructor. I can probably clean this up a bit more, but this is very helpful... thanks!

On my system, the copy constructor is getting called 12 times. Could you please explain what is this number related to?

```
#include <vector>
#include <iostream>
#include <numeric>

#include "Kokkos_Core.hpp"

int const size = 1e7;

void testSerial()
{
  std::vector<float> A(size,0.0);
  std::vector<float> B(size,0.0);
  std::vector<float> C(size,0.0);

  for (int i = 0; i < size; ++i)
  {
    A[i] = log(i+1);
    B[i] = log(i+1);
  }

  Kokkos::Timer kTimer;
  for (int i = 0; i < size; ++i)
  {
    C[i] = A[i]*B[i];
  }
  float const time = kTimer.seconds();

  float const result = std::accumulate(C.begin(), C.end(), 0.0);

  std::cout<<"testSerial: result = " << result << " in time = " << time << "(s)" << std::endl;
}

void testDirectKokkos()
{
  std::vector<float> A(size,0.0);
  std::vector<float> B(size,0.0);
  std::vector<float> C(size,0.0);

  for (int i = 0; i < size; ++i)
  {
    A[i] = log(i+1);
    B[i] = log(i+1);
  }

  typedef Kokkos::View<float*,Kokkos::Cuda> GpuView;
  GpuView gpuA("gpuA", size);
  GpuView gpuB("gpuB", size);
  GpuView gpuC("gpuC", size);

  typedef Kokkos::View<float*, Kokkos::HostSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>> UnmanagedHostView;
  Kokkos::deep_copy(gpuA, UnmanagedHostView(A.data(), A.size()));
  Kokkos::deep_copy(gpuB, UnmanagedHostView(B.data(), B.size()));

  Kokkos::Timer kTimer;
  Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::Cuda>(0,size), KOKKOS_LAMBDA(int i)
  {
    gpuC(i) = gpuA(i)*gpuB(i);
  });
  float const time = kTimer.seconds();

  Kokkos::deep_copy(UnmanagedHostView(C.data(), C.size()), gpuC);
  float const result = std::accumulate(C.begin(), C.end(), 0.0);

  std::cout<<"testDirectKokkos: result = " << result << " in time = " << time << "(s)" << std::endl;
}

// A wrapper class for the Kokkos data
class ConstKokkosField
{
public:
  typedef Kokkos::View<float*,Kokkos::HostSpace,Kokkos::MemoryTraits<Kokkos::Unmanaged>> UnmanagedHostView;

  ConstKokkosField(std::vector<float> &vec)
    : _size(vec.size())
    , _gpuField("_gpuField", _size)
    , _hostX(UnmanagedHostView(vec.data(), vec.size()))
  {
    Kokkos::deep_copy(_gpuField, _hostX);
  }

  ~ConstKokkosField() {}

  KOKKOS_FORCEINLINE_FUNCTION float& operator[](int const i) const
  {
    return _gpuField(i);
  }

protected:
  int const _size;
  typedef Kokkos::View<float*,Kokkos::Cuda> GpuView;
  GpuView _gpuField;
  typename GpuView::HostMirror _hostX;
};

class KokkosField
{
public:
  typedef Kokkos::View<float*,Kokkos::HostSpace,Kokkos::MemoryTraits<Kokkos::Unmanaged>> UnmanagedHostView;

  KokkosField(std::vector<float> &vec)
    : _size(vec.size())
    , _gpuField("_gpuField", _size)
    , _hostX(UnmanagedHostView(vec.data(), vec.size()))
    , _copyData(true)
  {
    //std::cout<<"calling KokkosField ctor"<<std::endl;
    Kokkos::deep_copy(_gpuField, _hostX);
  }

  KokkosField(KokkosField const& rhs)
    : _size(rhs._size)
    , _gpuField(rhs._gpuField)
    , _hostX(rhs._hostX)
    , _copyData(false)
  {
    //std::cout<<"calling KokkosField copy ctor"<<std::endl;
  }

  ~KokkosField() 
  {
    //std::cout<<"calling ~KokkosField dtor" << std::endl;
    if (_copyData)
      Kokkos::deep_copy(_hostX, _gpuField);
  }

  KOKKOS_FORCEINLINE_FUNCTION float& operator[](int const i) const
  {
    return _gpuField(i);
  }

protected:
  int const _size;
  typedef Kokkos::View<float*,Kokkos::Cuda> GpuView;
  GpuView _gpuField;
  typename GpuView::HostMirror _hostX;
  bool _copyData;
};

void testKokkosField()
{
  std::vector<float> A(size,0.0);
  std::vector<float> B(size,0.0);
  std::vector<float> C(size,0.0);

  for (int i = 0; i < size; ++i)
  {
    A[i] = log(i+1);
    B[i] = log(i+1);
  }

  float time;
  {
    ConstKokkosField kA(A);
    ConstKokkosField kB(B);
    KokkosField kC(C);

    Kokkos::Timer kTimer;
    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::Cuda>(0,size), KOKKOS_LAMBDA(int i)
    {
      kC[i] = kA[i]*kB[i];
    });
    time = kTimer.seconds();
  }

  float const result = std::accumulate(C.begin(), C.end(), 0.0);
  std::cout<<"testKokkosField: result = " << result << " in time = " << time << "(s)" << std::endl;
}

int main()
{
  Kokkos::InitArguments args;
  args.num_threads = 8;
  args.num_numa    = 2;
  args.device_id   = 0;
  Kokkos::initialize(args);

  testSerial();
  testDirectKokkos();
  testKokkosField();

  Kokkos::finalize();
  return 1;
}
```

**Comment by mhoemmen** [comment]
Kokkos must capture functors and lambdas by value in general.  It needs to do that in order to run correctly on the GPU.  For example, if Kokkos captured some type T by reference, that would be a reference to host data.  It would be incorrect to access that on host.

"Capture by value" means "copy."  That's why your functor's copy constructor gets invoked.

This is why Kokkos::View's copy constructor and copy assignment operator do shallow copies (behaving like a pointer), not deep copies (as std::vector does).  The idea is that your functors and lambdas should represent large amounts of data using Kokkos::View.

In short: capture pointers and Kokkos::View, not containers like std::vector.

**Comment by dalg24** [comment]
How many times the copy constructor will be called depends on the backend and you should see it as implementation details.

In any case, the first copy occurs in the Kokkos lambda capture clause.
Your lambda is passed to the specialization of `Kokkos::Impl::ParallelFor` that matches the execution policy you passed to the `parallel_for()`.  The current implementation of the CUDA "one-D" range policy specialization makes [another copy](https://github.com/kokkos/kokkos/blob/48deb6302b71ff6d3ecefe5767475a3388e3eb09/core/src/Cuda/Kokkos_Cuda_Parallel.hpp#L620) (so does the [serial backend](https://github.com/kokkos/kokkos/blob/48deb6302b71ff6d3ecefe5767475a3388e3eb09/core/src/Kokkos_Serial.hpp#L548)), which in turn is passed (by reference) to `CudaParallelLaunch` and then ... it depends :)

**Comment by varunhiremath** [comment]
Thank you very much for the explanation!
